### ISSUE TITLE
Clicking on code inside a decorator doesn't highlight that part of the AST

### ISSUE THREAD

**Issue by jseminck** [fkling/astexplorer#269]
Hello,

FIrst of all thanks so much for ASTExplorer. It's such a great tool and as an eslint-plugin developer I use it a lot and appreciate the work.

I recently ran into an issue that I want to write some eslint rules for a home-grown library that we use solely within decorators (kind of like @connect). But I found out that selecting part of the code inside the decorator doesn't highlight the relevant part in the AST like it does with normal code.

Here is an example:

https://astexplorer.net/#/gist/1733789f1377e41502e844b2f9ac5a6d/87e23677115df42e7a3a1398b0ef609b6869daaf

* When I click "Test" from "Class Test extends Component" then the part of the AST gets highlighted
* When I click on "actions.finish$" in the decorator, then the AST doesn't get highlighted and jumps to the top

I'd love to help out with fixing this if it's not too hard 😄  If you can give me some pointers where to look then I'd gladly help.

**Comment by jseminck** [comment]
I think the reason is that `getFocusPath` exits too early.

It checks the range of the `ExportDefaultDeclaration` and decides that, because the current cursor is not within that range (it is not, it's smaller than the range), then we don't have to recursively visit it's children.

However, the decorator is only defined 2 children deep, and that range is smaller than the range defined on the `ExportDefaultDeclaration` ...

Any ideas how to fix this?

<img width="628" alt="screen shot 2017-08-31 at 21 42 30" src="https://user-images.githubusercontent.com/9586897/29939688-5e9bc3c6-8e95-11e7-8fab-3fb5bea23d37.png">

**Comment by fkling** [comment]
Ah good catch! Yeah, the assumption was that a parent's range will always be larger or equal than its children's ranges. That doesn't work for decorators.

Not sure what the best solution is here. If we assume that a position could be contained inside any node, then we would have to traverse/find every node, which wouldn't be very efficient I think.

**Comment by jseminck** [comment]
I upgraded babylon7 to the latest beta. It updates the range of the `ClassExpression`, but not of the `DefaultExportExpression`. Not sure yet if this is intended or not. Will find out from the babel team first.

<img width="431" alt="screen shot 2017-08-31 at 22 23 15" src="https://user-images.githubusercontent.com/9586897/29941491-705b9d1a-8e9b-11e7-878d-a2a04c27e1e4.png">

**Comment by RReverser** [comment]
@fkling we can cache range-node maps, that shouldn't be a big problem.

**Comment by jseminck** [comment]
That means traversing through the whole node tree on the first "click", and caching everything. And then when the code changes, the cache has to be cleared. Or did I interpret it wrong?

**Comment by RReverser** [comment]
Yup, or traversing on parsing. Either way, that can turn out to be even faster than what we are doing right for certain cases.

**Comment by jseminck** [comment]
I might have some time this weekend to look into it!

**Comment by jseminck** [comment]
I managed to create an alternative `getFocusPath` function that seems to return a valid path. However, the path starts with the "Decorator" node, so the "ExportDefault" and "ClassDeclaration" nodes are being left out 😞 

I suppose the app needs the full path (starting from Program), but in the decorator case even the Program start doesn't include the decorator range.

<img width="692" alt="screen shot 2017-09-09 at 17 25 12" src="https://user-images.githubusercontent.com/9586897/30241010-efeb872c-9583-11e7-8375-ec420ed71656.png">

<img width="914" alt="screen shot 2017-09-09 at 17 22 33" src="https://user-images.githubusercontent.com/9586897/30240985-a1f88290-9583-11e7-9eba-039159421b65.png">
